### PR TITLE
viscous dynamics

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
The changes are summarized as follows:
1.  fixed a bug is old royld b case
2. [add shear dependent viscosity](https://github.com/Xiangyu-Hu/SPHinXsys/commit/c0812c163a8ed3b0b201ceec9bf4871f36ee1d17)
3. [add observer using lattice generator to lid driven cavity](https://github.com/Xiangyu-Hu/SPHinXsys/commit/824572bd880e1f33f340af14256b6564f52973d8)
4.  introduced signed distance to solid particles, and DistanceFromWall class for fluid particles.
5. Near linear reproducing boundary condition is added to velocity gradient computation and angular momentum conservative viscous force and a new test 2d_velocity_gradient is added as a unit test.
6. 2d lid driven as new case for future clarification of shear-dependent viscous flow test , but now regression test is not done yet.

### CODE REVIEW

File: src/shared/materials/weakly_compressible_fluid.cpp [hunk 1]
```diff
@@ -40,5 +40,20 @@ Real SymmetricTaitFluid::getSoundSpeed(Real p, Real rho)
                ? sqrt((p0_ + Real(gamma_) * p) / rho)
                : sqrt((p0_ - Real(gamma_) * p) / rho);
 }
+//=================================================================================================//
+Real HerschelBulkleyFluid::getViscosity(Real shear_rate)
+{
+
+    Real effective_shear_rate = SMAX(SMIN(shear_rate, max_shear_rate_), min_shear_rate_);
```
Comment (FabienPean-Virtonomy) [inline src/shared/materials/weakly_compressible_fluid.cpp:47]:
```suggestion
    Real effective_shear_rate = std::clamp(shear_rate, min_shear_rate_, max_shear_rate_);
```

File: src/shared/materials/weakly_compressible_fluid.cpp [hunk 2]
```diff
@@ -40,5 +40,20 @@ Real SymmetricTaitFluid::getSoundSpeed(Real p, Real rho)
...
+    return (yield_stress_ + consistency_index_ * (std::pow(effective_shear_rate, power_index_))) /
+           effective_shear_rate;
+}
+//=================================================================================================//
+Real CarreauFluid::getViscosity(Real shear_rate)
+{
+    Real effective_shear_rate = SMAX(SMIN(shear_rate, max_shear_rate_), min_shear_rate_);
```
Comment (FabienPean-Virtonomy) [inline src/shared/materials/weakly_compressible_fluid.cpp:54]:
```suggestion
    Real effective_shear_rate = std::clamp(shear_rate, min_shear_rate_, max_shear_rate_);
```

Apply similarly if it occurs elsewhere

File: src/shared/particle_dynamics/continuum_dynamics/continuum_integration.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@ ShearStressRelaxation ::
       velocity_gradient_(particles_->velocity_gradient_), strain_tensor_(particles_->strain_tensor_),
       strain_tensor_rate_(particles_->strain_tensor_rate_), von_mises_stress_(particles_->von_mises_stress_),
       von_mises_strain_(particles_->von_mises_strain_), Vol_(particles_->Vol_),
-      B_(*particles_->getVariableByName<Matd>("KernelCorrectionMatrix")) {}
+      B_(*particles_->getVariableByName<Matd>("FirstConsistencyMatrix")) {}
```
Comment (FabienPean-Virtonomy) [inline src/shared/particle_dynamics/continuum_dynamics/continuum_integration.cpp:43]: If the name changes, might as well call go  with the full name `FirstOrderConsistencyCorrectionMatrix`

File: src/shared/particle_dynamics/fluid_dynamics/fluid_integration.hpp [hunk 4]
```diff
@@ -71,7 +71,7 @@ void Integration1stHalf<Inner<>, RiemannSolverType, KernelCorrectionType>::inter
         Real dW_ijV_j = inner_neighborhood.dW_ijV_j_[n];
         const Vecd &e_ij = inner_neighborhood.e_ij_[n];
 
-        force -= mass_[index_i] * (p_[index_i] * correction_(index_i) + p_[index_j] * correction_(index_j)) * dW_ijV_j * e_ij;
+        force -= mass_[index_i] * (p_[index_i] * correction_(index_j) + p_[index_j] * correction_(index_i)) * dW_ijV_j * e_ij;
```
Comment (FabienPean-Virtonomy) [inline src/shared/particle_dynamics/fluid_dynamics/fluid_integration.hpp:74]: @WeiyiVirtonomy Should we backport those changes ?

Comment (Xiangyu-Hu) [reply src/shared/particle_dynamics/fluid_dynamics/fluid_integration.hpp:74]: Sorry. I will chnage it back. This form will be tested latter in another branch.

File: src/shared/particle_dynamics/continuum_dynamics/continuum_integration.cpp (hunk 3 above)
Comment (Xiangyu-Hu) [reply src/shared/particle_dynamics/continuum_dynamics/continuum_integration.cpp]: I have renamed it to LinearGradientCorrectionMatrix for more expressive naming

File: src/shared/materials/weakly_compressible_fluid.cpp (hunk 1 above)
Comment (Xiangyu-Hu) [reply src/shared/materials/weakly_compressible_fluid.cpp:47]: I will change it in the next pull request.

**Comment by Xiangyu-Hu** [comment]
@FabienPean-Virtonomy @FengWang3119 @TheodorH999 @Bo-Zhang1995 @Lisu-lisa Please have check the pull request. If there is no issues, please approve it. Thanks.

**Review by Lisu-lisa — APPROVED** [review]
